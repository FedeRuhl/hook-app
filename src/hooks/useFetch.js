import { useEffect, useRef, useState } from "react";

const useFetch = (url) => {

    const isMounted = useRef(true);

    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null
    });

    useEffect(() => {
        
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {

        setState({
            data: null,
            loading: true,
            error: null
        });

        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                if (isMounted.current)
                {
                    setState({
                        data,
                        loading: false,
                        error: null
                    });
                }
            })
            .catch((e) => setState({
                data: null,
                loading: false,
                error: 'Data could not be fetched'
            }));
    }, [url]);

    return state;

};

export default useFetch;