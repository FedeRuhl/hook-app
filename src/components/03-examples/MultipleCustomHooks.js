import React from 'react';
import useFetch from '../../hooks/useFetch';
import useCounter from '../../hooks/useCounter';
import './effects.css';

const MultipleCustomHooks = () => {

    const { state: counter, increment } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

    const author = (!!data && data.length > 0) ? data[0].author : '';
    const quote = (!!data && data.length > 0) ? data[0].quote : '';

    return (
        <>
        <h1>
            Breaking Bad Quotes
        </h1>

        <hr />

        { 
        
        (loading)
            ? 
                (
                    <div className="alert alert-info text-center">Cargando...</div>
                )
            :
                (
                    <blockquote className="blockquote text-end">
                        <p className="mb-2">{ quote }</p>
                        <footer className="blockquote-footer">{ author }</footer>
                    </blockquote>
                )
        }

        <button className="btn btn-primary" onClick={ () => increment() }>
            Siguiente cita
        </button>

        </>
    )
}

export default MultipleCustomHooks
