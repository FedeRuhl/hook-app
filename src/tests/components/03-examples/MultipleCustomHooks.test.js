import React from 'react';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import useFetch from '../../../hooks/useFetch';
import useCounter from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('MultipleCustomHooks tests', () => {

    beforeEach(() => {
        useCounter.mockReturnValue({
            counter: 10,
            increment: () => {}
        });
    });

    test('should show the component succesfully', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should show the info', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Fernando',
                quote: 'Hola mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-2').text().trim()).toBe('Hola mundo');
        expect(wrapper.find('footer').text().trim()).toBe('Fernando');

    });
    
});