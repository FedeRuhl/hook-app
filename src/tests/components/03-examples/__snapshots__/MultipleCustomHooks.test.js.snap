// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks tests should show the component succesfully 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Cargando...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente cita
  </button>
</Fragment>
`;
